### PR TITLE
Fix Single ThresholdBug in Simple Blob Detector

### PR DESCRIPTION
**Pull request by danielenricocahall**
Per the bug mentioned in https://github.com/opencv/opencv/issues/6667, if there is only one threshold, the logic which uses `minDistBetweenBlobs` is not reached. This PR addresses that issue by ensuring that, if there is only one threshold, we add `curCenters` before the blob distance checking logic. The test replicates the test setup described in the issue, which originally failed but now passes, as `centers` is now populated and therefore we extract keypoints. 

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/features2d/test/test_blobdetector.cpp [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#include "test_precomp.hpp"
+
+namespace opencv_test {
+    TEST(Features2d_BlobDetector, bug_6667) {
```
Comment (alalek) [inline modules/features2d/test/test_blobdetector.cpp:4]: Avoid indentation in namespaces.

File: modules/features2d/test/test_blobdetector.cpp [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#include "test_precomp.hpp"
```
Comment (alalek) [inline modules/features2d/test/test_blobdetector.cpp:1]: Please use short license header: https://github.com/opencv/opencv/wiki/Coding_Style_Guide#file-structure

```
// This file is part of OpenCV project.
// It is subject to the license terms in the LICENSE file found in the top-level directory
// of this distribution and at http://opencv.org/license.html.
```

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

File: modules/features2d/test/test_blobdetector.cpp (hunk 1 above)
Comment (danielenricocahall) [reply modules/features2d/test/test_blobdetector.cpp]: Resolved!

File: modules/features2d/test/test_blobdetector.cpp (hunk 2 above)
Comment (danielenricocahall) [reply modules/features2d/test/test_blobdetector.cpp:5]: Resolved!

**Review by alalek — APPROVED** [review]
Thank you :+1:

File: modules/features2d/src/blobdetector.cpp [hunk 3]
```diff
@@ -325,13 +325,19 @@ void SimpleBlobDetectorImpl::detect(InputArray image, std::vector<cv::KeyPoint>&
 
         std::vector < Center > curCenters;
         findBlobs(grayscaleImage, binarizedImage, curCenters);
+        if(params.maxThreshold - params.minThreshold <= params.thresholdStep) {
+            // if the difference between min and max threshold is less than the threshold step
+            // we're only going to enter the loop once, so we need to add curCenters
+            // to ensure we still use minDistBetweenBlobs
+            centers.push_back(curCenters);
```
Comment (alalek) [inline modules/features2d/src/blobdetector.cpp:332]: This change is considered as invalid.
It produces invalid keypoint with mean point of centers ([code](https://github.com/opencv/opencv/blob/3.4.15/modules/features2d/src/blobdetector.cpp#L367-L374))

relates #20649

File: modules/features2d/test/test_blobdetector.cpp [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace opencv_test { namespace {
+TEST(Features2d_BlobDetector, bug_6667)
+{
+    cv::Mat image = cv::Mat(cv::Size(100, 100), CV_8UC1, cv::Scalar(255, 255, 255));
+    cv::circle(image, Point(50, 50), 20, cv::Scalar(0), -1);
+    SimpleBlobDetector::Params params;
+    params.minThreshold = 250;
+    params.maxThreshold = 260;
+    std::vector<KeyPoint> keypoints;
```
Comment (alalek) [inline modules/features2d/test/test_blobdetector.cpp:15]: `params.minRepeatability = 2` by default.
This is why no keypoints are detected here if there is one "threshold" only (which parameters combination doesn't makes sense).

**Pull request merged**